.github/workflows/release.yml

```diff
@@ -98,12 +98,171 @@ jobs:
+        with:
+          name: layer.tf
+          path: layer.tf
+  smoke-test:
+    runs-on: ubuntu-latest
+    needs: generate-note
+    name: Smoke Test - ${{ matrix.aws_region }} - ${{ matrix.name }}
+    strategy:
+      fail-fast: false
+      matrix:
+        aws_region: ${{fromJson(github.event.inputs.aws_region).aws_region}}
+        # We need to define the raw parameters to the matrix here but will customize each
+        # below.
+        name:
+          - java-awssdk-agent
+          - java-awssdk-wrapper
+          - java-okhttp-wrapper
+          - nodejs-awssdk
+          - python38
+        include:
+          - name: java-awssdk-agent
+            layer_kind: java-agent
+            language: java
+            build_directory: java
+            build_command: ./build.sh
+            terraform_directory: sample-apps/java-agent-aws-sdk-terraform
+            expected_template: adot/utils/expected-templates/java-awssdk-agent.json
+          - name: java-awssdk-wrapper
+            layer_kind: java-wrapper
+            language: java
+            build_directory: java
+            build_command: ./build.sh
+            terraform_directory: sample-apps/java-wrapper-aws-sdk-terraform
+            expected_template: adot/utils/expected-templates/java-awssdk-wrapper.json
+          - name: java-okhttp-wrapper
+            layer_kind: java-wrapper
+            language: java
+            build_directory: java
+            build_command: ./build.sh
+            terraform_directory: sample-apps/java-wrapper-okhttp-terraform
+            expected_template: adot/utils/expected-templates/java-okhttp-wrapper.json
+          - name: nodejs-awssdk
+            layer_kind: nodejs
+            language: nodejs
+            build_directory: opentelemetry-lambda/nodejs
+            build_command: npm install
+            terraform_directory: sample-apps/nodejs-aws-sdk-terraform
+            expected_template: adot/utils/expected-templates/nodejs-awssdk.json
+          - name: python38
+            layer_kind: python38
+            language: python
+            build_directory: opentelemetry-lambda/python
+            build_command: |
+              cd sample-apps
+              ./build.sh
+            terraform_directory: sample-apps/python-aws-sdk-aiohttp-terraform
+            expected_template: adot/utils/expected-templates/python.json
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          submodules: true
+      - uses: actions/setup-java@v2
+        if: ${{ matrix.language == 'java' }}
+        with:
+          distribution: adopt
+          java-version: '11'
+      - name: Cache (Java)
+        uses: actions/cache@v2
+        if: ${{ matrix.language == 'java' }}
+        with:
+          path: |
+            ~/go/pkg/mod
+            ~/.gradle/caches
+            ~/.gradle/wrapper
+          key: ${{ runner.os }}-gradle-${{ hashFiles('**/*.gradle*', '**/gradle-wrapper.properties') }}-go-${{ hashFiles('**/go.sum') }}
```

Comment:
I'll leave it up to you, but maybe it would be cleaner if you could refactor this to have 1 step that always caches the go modules (regardless of `matrix.language`), and then these conditional caches would only apply to that language's artifacts.